nixos/iso-image: Consolidate ISOs Using Specialisations.

ElvishJerricco opened the pull request:
Not the most elegant of code just yet, so marking as draft for now.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ElvishJerricco committed:
nixos/iso-image: Include latest kernel specialisation.

Now that ISOs include the latest kernel as a specialisation, separate
ISOs for the new kernel are no longer necessary.

ElvishJerricco committed:
nixos/iso-image: Disable plymouth along with display-manager.service

GNOME in particular just breaks if plymouth isn't disabled, because
GDM takes on the role of quitting plymouth in a GNOME
configuration. But if we're disabling the DM, we should disable
plymouth too anyway.

ElvishJerricco committed:
nixos/iso-image: Remove video= boot entries

These seem largely redundant with the progress that desktop
environments have made since they were introduced.

ElvishJerricco committed:
nixos/iso-image: Support specialisations.

ElvishJerricco committed:
nixos-generate-config: Set linuxPackages_latest when ISO boots latest kernel

ElvishJerricco committed:
nixos-generate-config: Add config file

ElvishJerricco commented:
@vlinkz My only remaining concern with this PR before marking ready for review is that I don't think the graphical installer will automatically select the latest kernel when that specialisation is booted. Do you have any ideas for how to fix that?

vlinkz commented:
> @vlinkz My only remaining concern with this PR before marking ready for review is that I don't think the graphical installer will automatically select the latest kernel when that specialisation is booted. Do you have any ideas for how to fix that?

`/etc/nixos-generate-config.conf` has the specialisation config for the booted session right? We can just parse that file and add the option as needed

vlinkz commented:
Not sure how far you're planning on upgrading `nixos-generate-config`, but another option could be add all the options currently supported by the graphical installer to `nixos-generate-config` (like `--desktop=gnome/kde`) and then use `nixos-generate-config` within calamares instead of the current custom snippets.

ElvishJerricco commented:
@vlinkz Yea, I think unifying `nixos-generate-config` with the calamares installer would be a great step. Heck, we could rewrite `nixos-generate-config` in python or rust or something (to go with all the other `-ng` rewrites lately) and add options to this new `nixos-generate-config.conf` file so that calamares can just generate that file instead of generating nix expressions.

(Also, I should probably make it TOML instead of INI)

ElvishJerricco commented:
That said, I *hadn't* originally been planning to upgrade `nixos-generate-config` much at all. So maybe to start with we just leave it the way it is (except TOML) and parse the kernel option in calamares. We can make it fancy later.

vlinkz commented:
> Heck, we could rewrite `nixos-generate-config` in python or rust or something

Definitely a fan of that idea! But yeah definitely a longer term project.

> That said, I _hadn't_ originally been planning to upgrade `nixos-generate-config` much at all. So maybe to start with we just leave it the way it is (except TOML) and parse the kernel option in calamares. We can make it fancy later.

Sounds good, I'll work on a way for calamares to detect the file and modify the config accordingly.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5062

ElvishJerricco committed:
iso-image: Combine GNOME and Plasma flavors

This allows users to choose between the two during bootup rather than
at download time.

ElvishJerricco committed:
nixos/iso-image.nix: nixfmt

ElvishJerricco committed:
.git-blame-ignore-revs: Add iso-image.nix commit

marcusramberg commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -68,11 +75,19 @@ sub runCommand {
+    elsif ($arg eq "--kernel") {
+        $n++;
+        $kernel = $ARGV[$n];
+        die "$0: ‘--kernel’ requires an argument\n" unless defined $kernel;
```
Maybe you should also handle the case where $kernel starts with -- ?
```suggestion
        die "$0: ‘--kernel’ requires an argument\n" unless defined $kernel && $kernel !~ /^--/;
```

ElvishJerricco replied:
@marcusramberg That's (effectively) already handled below, where the value of the `--kernel` argument is checked for validity. I would think the error `invalid kernel: '--root'` would be clear enough

ElvishJerricco committed:
nixos/iso-image: Move alternate boot options to submenus

This will make more sense when specialisations are introduced in the
ISO boot menu

nixos/modules/installer/cd-dvd/iso-image.nix

mweinelt reviewed (approved):
I gave this a spin in a virtual machine, and things are working as expected.

Also, I skimmed the code and it looks plausible.

mweinelt reviewed (approved):
Tested and still looking fine.

At some point down the road we should probably look into making the options toggles, so that you can check multiple of them, if that is at all possible.

ElvishJerricco commented:
> At some point down the road we should probably look into making the options toggles, so that you can check multiple of them, if that is at all possible.

If it's possible, I don't know how to do it :P

spiage commented:
@ElvishJerricco @vlinkz @marcusramberg @mweinelt Hi!
Can you please help me to understand why nixos-generate-config is in my system-path?
As I can see it is a part of installer...

spiage commented:
I hope it can help
```
❯ nix-store --query --referrers-closure /nix/store/i42n7r4plk72wj5bnv2ccnmbgnpql0j4-nixos-generate-config/
/nix/store/i42n7r4plk72wj5bnv2ccnmbgnpql0j4-nixos-generate-config
/nix/store/c9fhs7bqkachblcryiaip10q35vszz4k-system-path
/nix/store/vzmxb69dnzf0kb7psik5279b25f9a6jc-X-Restart-Triggers-polkit
/nix/store/19bm23g7qr6c21vf95ghqlq3hd677449-unit-polkit.service
/nix/store/gc3pr3d8022xpcb0kfbs3hvzsv39i9z6-dbus-1
/nix/store/590qvh42a7q6ipkih69m1rj3qikf0mnq-X-Restart-Triggers-dbus
/nix/store/xr4v9296n5sk9g45n8gji2w4kq6pxwii-unit-dbus.service
/nix/store/pgw1q4zg2r4s5bg42vz5l5r0s6nvbb4v-user-units
/nix/store/87zfz49g4c9pfqr799a1qi7rb48xcp85-unit-dbus.service
/nix/store/y01dhpw5i6paai1jlbw16crp12yw06jr-unit-accounts-daemon.service
/nix/store/zmwq9dgp59czbc23gizb671ihg4wyif7-system-units
/nix/store/a48f3wmnlwv30dqlvsv9da895i1v1z73-etc
/nix/store/4hij6p3wfd1f1jdhbir0xv7n3yaj8wsm-nixos-system-a7-25.05.20250318.317736b

❯ nix-store --query --references /nix/store/i42n7r4plk72wj5bnv2ccnmbgnpql0j4-nixos-generate-config/
/nix/store/57rayvb28n66prgyavcbz2kk1qv3zydw-systemd-257.3
/nix/store/grcrzm31kv0y73q1p565niaaca5dn2xp-perl-5.40.0-env
/nix/store/ssw00442c7waswgxwhmd2wg7vzwhx2sj-btrfs-progs-6.13
```

ElvishJerricco commented:
@spiage Please use the appropriate forums and chatrooms for support questions. It is normal for `nixos-generate-config` to be on the system-path.

spiage commented:
@ElvishJerricco  Sorry
I understand now that I can use this tool to create config for running system, not only while installing =)
And I'll use discourse for questions next time

vcunat commented:
@raboof: I suspect that we've forgotten that some changes are needed in the **reproducibility**-checking infra?  There's no `iso_gnome` anymore; the job is called `iso_graphical` instead (containing both gnome and KDE/Plasma).

raboof commented:
Thanks for the heads-up, indeed I hadn't noticed yet (I've been focused on the minimal ISO and on https://github.com/JulienMalka/lila - likely more activity there soon!). I'll get things going for the graphical ISO instead of the separate gnome/plasma ones.

vcunat commented:
If you work with the `nixos-unstable` channel, this change only reached it several minutes ago 😉

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/a-few-annecdotal-nixos-install-ux-issues/63546/12
